File: internal/actions/actions.go
// Package actions holds the backend rules for which actions a PR's menu offers.
// Clients display the results; they never decide availability themselves.
package actions

import (
	"fmt"
	"strings"

	"github.com/acheris-labs/pr-mon/internal/models"
	"github.com/acheris-labs/pr-mon/internal/readiness"
)

const (
	noMethods   = "no merge methods allowed"
	keepsBranch = "branch won't be deleted: repo doesn't auto-delete"
	afterWaits  = "after the PRs it waits on have merged"
)

// WithActions returns the repo with every PR's action menu filled in.
func WithActions(repo models.Repo, armed map[int]models.ArmedMerge) models.Repo {
	prs := make([]models.PullRequest, len(repo.PRs))
	for i, pr := range repo.PRs {
		pr.Actions = Available(repo, pr, armedFor(armed, pr.Number))
		prs[i] = pr
	}
	repo.PRs = prs
	return repo
}

func armedFor(armed map[int]models.ArmedMerge, number int) *models.ArmedMerge {
	if merge, ok := armed[number]; ok {
		return &merge
	}
	return nil
}

func Available(repo models.Repo, pr models.PullRequest, armed *models.ArmedMerge) []models.ActionOption {
	deletable := !repo.DeleteBranchOnMerge && pr.HeadRefID != nil
	options := []models.ActionOption{
		mergeOption(repo, pr, deletable),
		autoMergeOption(repo, pr, armed, deletable),
	}
	if pr.MergeState == "BEHIND" {
		options = append(options, models.ActionOption{
			Key: "update", Kind: "update", Label: "Update branch", Available: true,
		})
	}
	options = append(options, draftOption(pr))
	if rerun := rerunOption(pr); rerun != nil {
		options = append(options, *rerun)
	}
	return options
}

// rerunOption re-runs the failed jobs of this PR's checks; nil when GitHub
// Actions didn't run any (another CI system, or nothing has run yet).
func rerunOption(pr models.PullRequest) *models.ActionOption {
	actionsRan, failed := false, false
	for _, check := range pr.Checks {
		if check.RunID == nil {
			continue
		}
		actionsRan = true
		failed = failed || readiness.CheckFailed(check)
	}
	if !actionsRan {
		return nil
	}
	if !failed {
		return &models.ActionOption{
			Key: "rerun", Kind: "rerun_checks", Label: "Re-run failed checks",
			Available: false, Reason: models.Ptr("no failed checks"),
		}
	}
	return &models.ActionOption{
		Key: "rerun", Kind: "rerun_checks", Label: "Re-run failed checks", Available: true,
	}
}

// draftOption takes a PR out of draft, or puts it back.
func draftOption(pr models.PullRequest) models.ActionOption {
	if pr.IsDraft {
		return models.ActionOption{
			Key: "draft", Kind: "mark_ready", Label: "Mark ready for review", Available: true,
		}
	}
	return models.ActionOption{
		Key: "draft", Kind: "convert_to_draft", Label: "Convert to draft", Available: true,
	}
}

func mergeOption(repo models.Repo, pr models.PullRequest, deletable bool) models.ActionOption {
	if pr.Status == models.StatusReady && len(repo.MergeMethods) > 0 {
		return models.ActionOption{
			Key: "merge", Kind: "merge", Label: "Merge", Available: true,
			NeedsMethod: true, OffersDeleteBranch: deletable,
		}
	}
	var why []string
	for _, reason := range pr.Reasons {
		if reason.Level != "info" {
			why = append(why, reason.Text)
		}
	}
	if len(repo.MergeMethods) == 0 {
		why = append(why, noMethods)
	}
	detail := string(pr.Status)
	if len(why) > 0 {
		detail = fmt.Sprintf("%s: %s", pr.Status, strings.Join(why, ", "))
	}
	return models.ActionOption{
		Key: "merge", Kind: "merge", Label: "Merge", Available: false, Reason: &detail,
	}
}

// autoMergeOption offers GitHub's native auto-merge where the repo allows it,
// else pr-mon's own merge when ready. A PR waiting on others always gets
// pr-mon's: GitHub would merge it without waiting.
func autoMergeOption(
	repo models.Repo, pr models.PullRequest, armed *models.ArmedMerge, deletable bool,
) models.ActionOption {
	if pr.AutoMerge != nil {
		return models.ActionOption{
			Key: "auto_merge", Kind: "auto_merge_off", Label: "Disable auto-merge", Available: true,
		}
	}
	if armed != nil {
		return models.ActionOption{
			Key: "auto_merge", Kind: "disarm_merge",
			Label: "Cancel merge when ready (pr-mon)", Available: true,
		}
	}
	native := repo.AutoMergeAllowed && !pr.Waiting()
	kind := "arm_merge"
	if native {
		kind = "auto_merge_on"
	}
	if blocker := autoMergeBlocker(repo, pr); blocker != "" {
		label := "Merge when ready"
		if native {
			label = "Auto-merge"
		}
		return models.ActionOption{
			Key: "auto_merge", Kind: kind, Label: label, Available: false, Reason: &blocker,
		}
	}
	if !native {
		option := models.ActionOption{
			Key: "auto_merge", Kind: kind, Label: "Merge when ready (pr-mon)", Available: true,
			NeedsMethod: true, OffersDeleteBranch: deletable,
		}
		if pr.Waiting() {
			option.Note = models.Ptr(afterWaits)
		}
		return option
	}
	option := models.ActionOption{
		Key: "auto_merge", Kind: kind, Label: "Enable auto-merge", Available: true,
		NeedsMethod: true,
	}
	if !repo.DeleteBranchOnMerge {
		option.Note = models.Ptr(keepsBranch)
	}
	return option
}

func autoMergeBlocker(repo models.Repo, pr models.PullRequest) string {
	switch {
	case len(repo.MergeMethods) == 0:
		return noMethods
	case pr.IsDraft:
		return "draft PR"
	case pr.Status == models.StatusReady:
		return "already mergeable"
	}
	return ""
}

File: src/pr_mon/actions.py
"""Backend rules: which actions a PR's menu offers. Clients display the results."""

from dataclasses import replace

from pr_mon.models import ActionOption, ArmedMerge, PullRequest, RepoInfo, Status

NO_METHODS = "no merge methods allowed"
KEEPS_BRANCH = "branch won't be deleted: repo doesn't auto-delete"


def with_actions(repo: RepoInfo, armed: dict[int, ArmedMerge]) -> RepoInfo:
    """The repo with every PR's action menu filled in."""
    prs = tuple(
        replace(pr, actions=available_actions(repo, pr, armed.get(pr.number))) for pr in repo.prs
    )
    return replace(repo, prs=prs)


def available_actions(
    repo: RepoInfo, pr: PullRequest, armed: ArmedMerge | None
) -> tuple[ActionOption, ...]:
    deletable = not repo.delete_branch_on_merge and bool(pr.head_ref_id)
    options = [_merge(repo, pr, deletable), _auto_merge(repo, pr, armed, deletable)]
    if pr.merge_state == "BEHIND":
        options.append(ActionOption("update", "update", "Update branch", True))
    return tuple(options)


def _merge(repo: RepoInfo, pr: PullRequest, deletable: bool) -> ActionOption:
    if pr.status == Status.READY and repo.merge_methods:
        return ActionOption(
            "merge", "merge", "Merge", True, needs_method=True, offers_delete_branch=deletable
        )
    why = [reason.text for reason in pr.reasons if reason.level != "info"]
    if not repo.merge_methods:
        why.append(NO_METHODS)
    detail = f"{pr.status}: {', '.join(why)}" if why else str(pr.status)
    return ActionOption("merge", "merge", "Merge", False, reason=detail)


def _auto_merge(
    repo: RepoInfo, pr: PullRequest, armed: ArmedMerge | None, deletable: bool
) -> ActionOption:
    """GitHub's native auto-merge where the repo allows it, else pr-mon's merge when ready."""
    if pr.auto_merge:
        return ActionOption("auto_merge", "auto_merge_off", "Disable auto-merge", True)
    if armed:
        return ActionOption("auto_merge", "disarm_merge", "Cancel merge when ready (pr-mon)", True)
    native = repo.auto_merge_allowed
    kind = "auto_merge_on" if native else "arm_merge"
    blocker = _auto_merge_blocker(repo, pr)
    if blocker:
        label = "Auto-merge" if native else "Merge when ready"
        return ActionOption("auto_merge", kind, label, False, reason=blocker)
    if not native:
        return ActionOption(
            "auto_merge",
            kind,
            "Merge when ready (pr-mon)",
            True,
            needs_method=True,
            offers_delete_branch=deletable,
        )
    note = (
        None if repo.delete_branch_on_merge else "branch won't be deleted: repo doesn't auto-delete"
    )
    return ActionOption("auto_merge", kind, "Enable auto-merge", True, note=note, needs_method=True)


def _auto_merge_blocker(repo: RepoInfo, pr: PullRequest) -> str | None:
    if not repo.merge_methods:
        return NO_METHODS
    if pr.is_draft:
        return "draft PR"
    if pr.status == Status.READY:
        return "already mergeable — use m"
    return None
